rust/kernel/sync/arc.rs

```diff
@@ -122,24 +128,71 @@ impl<T> Ref<T> {
     /// `encoded` must have been returned by a previous call to [`Ref::into_usize`]. Additionally,
     /// it can only be called once for each previous call to [``Ref::into_usize`].
     pub unsafe fn from_usize(encoded: usize) -> Self {
+        // SAFETY: By the safety invariants we know that `encoded` came from `Ref::into_usize`, so
+        // the reference count held then will be owned by the new `Ref` object.
+        unsafe { Self::from_inner(NonNull::new(encoded as _).unwrap()) }
+    }
+}
+
+impl<T: ?Sized> Ref<T> {
+    /// Constructs a new [`Ref`] from an existing [`RefInner`].
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that `inner` points to a valid location and has a non-zero reference
+    /// count, one of which will be owned by the new [`Ref`] instance.
+    unsafe fn from_inner(inner: NonNull<RefInner<T>>) -> Self {
+        // INVARIANT: By the safety requirements, the invariants hold.
         Ref {
-            ptr: NonNull::new(encoded as _).unwrap(),
+            ptr: inner,
             _p: PhantomData,
         }
     }
-}
 
-impl<T: ?Sized> Ref<T> {
     /// Determines if two reference-counted pointers point to the same underlying instance of `T`.
     pub fn ptr_eq(a: &Self, b: &Self) -> bool {
-        core::ptr::eq(a.ptr.as_ptr(), b.ptr.as_ptr())
+        ptr::eq(a.ptr.as_ptr(), b.ptr.as_ptr())
     }
 
     /// Returns a pinned version of a given `Ref` instance.
     pub fn pinned(obj: Self) -> Pin<Self> {
         // SAFETY: The type invariants guarantee that the value is pinned.
         unsafe { Pin::new_unchecked(obj) }
     }
+
+    /// Deconstructs a [`Ref`] object into a raw pointer.
+    ///
+    /// It can be reconstructed once via [`Ref::from_raw`].
+    pub fn into_raw(obj: Self) -> *const T {
+        &**ManuallyDrop::new(obj)
```

Comment:
forget might be more readable than `ManuallyDrop`.

Reply:
Like this or do you have something better-looking?

```rs
        let ret: *const T = obj.deref();
        core::mem::forget(obj);
        ret
```

Reply:
I would personally write `let ret = &*obj as *const T`, but this looks equally good to me.

Reply:
Ok, I went with your suggestion.